tests/test_histogram.py

```diff
@@ -110,3 +117,85 @@ def test_compare_histograms_optimized_vs_general():
     assert_allclose(hist_ghc_root['sum_hessians'], hist_ghc_root['count'])
     assert_allclose(hist_gc['sum_hessians'], np.zeros(n_bins))  # unused
     assert_allclose(hist_ghc['sum_hessians'], hist_ghc['count'])
+
+
+def test_unrolled_equivalent_to_naive():
+    # Make sure the different unrolled histogram computations give the same
+    # results as the naive one.
+    rng = np.random.RandomState(42)
+    n_samples = 10
+    n_bins = 5
+    sample_indices = np.arange(n_samples).astype(np.uint32)
+    binned_feature = rng.randint(0, n_bins - 1, size=n_samples, dtype=np.uint8)
+    ordered_gradients = rng.randn(n_samples).astype(np.float32)
+    ordered_hessians = np.ones(n_samples, dtype=np.float32)
+
+    hist_gc_root = _build_gc_root_histogram_unrolled(n_bins, binned_feature,
+                                                     ordered_gradients)
+    hist_ghc_root = _build_ghc_root_histogram_unrolled(n_bins, binned_feature,
+                                                       ordered_gradients,
+                                                       ordered_hessians)
+    hist_gc = _build_gc_histogram_unrolled(n_bins, sample_indices,
+                                           binned_feature,
+                                           ordered_gradients)
+    hist_ghc = _build_ghc_histogram_unrolled(n_bins, sample_indices,
+                                             binned_feature,
+                                             ordered_gradients,
+                                             ordered_hessians)
+
+    hist_naive = _build_ghc_histogram_naive(n_bins, sample_indices,
+                                            binned_feature, ordered_gradients,
+                                            ordered_hessians)
```

Comment:
Because we use the unrolled we could rename all the `_build_*_histogram_unrolled` to just `_build_*_histogram` but keep the "naive" suffix in `_build_ghc_histogram_naive`.

Also the `gc` / `ghc` stuff is confusing.

Maybe:

- `_build_ghc_root_histogram_unrolled` =>`_build_histogram_root_node`;
- `_build_gc_root_histogram_unrolled` =>`_build_histogram_root_node_no_hessian`;
- `_build_ghc_histogram_unrolled` =>`_build_histogram`;
- `_build_gc_histogram_unrolled` =>`_build_histogram_no_hessian`.
- `_build_ghc_histogram_naive` => `_build_histogram_naive`